### ISSUE TITLE
Union same type and name requirement too strict

### ISSUE THREAD

**Issue by kuriboshi** [rbock/sqlpp11#124]
I'm trying to do a union between two different tables using a subset of columns from each table.  For example:

SELECT a.id FROM a UNION SELECT b.id FROM b

This seems to not be possible in sqlpp11 which will output the error "both arguments in a union have to have the same result columns (type and name)".  The following test program fails to compile with the above error: [sql.zip](https://github.com/rbock/sqlpp11/files/518352/sql.zip)

If I comment out the strict check in sqlpp11/union.h the program compiles and generates the expected SQL statement.  Maybe there is another way to accomplish this.

**Comment by rbock** [comment]
As both `a.id` and `b.id` have the same name and type (integral), that should work indeed.

I'll look into it asap.

**Comment by rbock** [comment]
For some reason, something seems to be wrong with the name analysis. For the time being, you can help yourself with:

```
auto u = sqlpp::select(aa.id).from(aa).unconditionally().union_distinct(
                  sqlpp::select(bb.id.as(aa.id)).from(bb).unconditionally());
```

**Comment by rbock** [comment]
The reason is clear now: result field specifications get their alias from the selected column. An alias is more than just the name. It is a unique type per column.

The solution is also clear:
When comparing result row types in a union, I need to compare the name only. 

I'll need some time to think about how to actually do that.

**Comment by kuriboshi** [comment]
Thank you. Your workaround solves the problem for the case when I select from a subset of columns from two tables.  However, I have similar problem when adding new columns which are either in neither table or in only one table.

I tried to solve it by declaring a new table which was a superset of the two tables I was making a union of and then using columns from this super table in the 'as' call.  I was getting the above static_assert error for some columns, not all of them, when one column was from a proper column in a table and the other was a constant string made in this way: sqlpp::value("test"s).as(super_table.test).

This case turned out to be because I had columns of type varchar and the two real tables I selected from where defined by code generated by the ddl2cpp script in the 'scripts' directory and the super table I created using SQLPP_DECLARE_TABLE from sqlpp11/ppgen.h.  It seems the code generated in these two different ways aren't always compatible with each other.  The solution for now is to stick to one or the other exclusively.

I'll try to come up with a small test case to demonstrate this behaviour.

**Comment by kuriboshi** [comment]
I created some test cases for the problem I mentioned in my previous comment: [sql2.zip](https://github.com/rbock/sqlpp11/files/519466/sql2.zip)

The problem is around the make_traits declaration.  If I use ppgen for all tables then I have to leave the third argument in the column declaration either blank or use SQLPP_NOT_NULL.  Using SQLPP_NULL doesn't work.  If I combine the ppgen tables with handcrafted column declarations then the make_traits declaration can not use sqlpp::tag::can_be_null.  In that case no variation of a column declaration in ppgen works.

Maybe my messy test cases don't provide much additional information to the issue at hand but I'll include them just in case.

**Comment by niXman** [comment]
Right now I have no free time to understand subj, so tell me what PPGEN doing wrong(if PPGEN doing it wrong), and I will fix it as soon as possible.

**Comment by kuriboshi** [comment]
I don't think ppgen is doing anything wrong.  The problem my second test case shows I think can be summarised in that I can't do a union of two tables with columns which allows null, given the other constraints of the issue (union of two different tables with different columns but with same column names/types).

**Comment by rbock** [comment]
Hi,

I pushed some changes to develop which hopefully go into the right direction. At least the old tests and your first sample compile fine now.

It definitely needs some more work for dynamic selects, but that will have to wait until tomorrow at least.

Best,

Roland

**Comment by rbock** [comment]
Unless you are using Windows, that is. I just saw that Appveyor (i.e. MSVC) is unhappy with the new code. Sigh...

**Comment by kuriboshi** [comment]
Thanks Roland.  I tried your changes in the develop branch (I'm using clang on OS X) and it does indeed resolve the issue I had originally.

**Comment by rbock** [comment]
Thanks for the feedback!

Are there remaining issues? To be honest, I did not really understand what I should do with your second example code.

**Comment by kuriboshi** [comment]
Sorry, my second case wasn't very clear since I wasn't sure what was causing the problem.  Here is a much reduced case which fails, even in the develop branch.  I'm doing a union of two tables where one of the tables is missing a column and I supply a default value for the missing a column.  The corresponding SQL is:

> SELECT a.id, a.other FROM a UNION SELECT b.id, "other" AS other FROM b

This fails with the same error ("both arguments in a union have to have the same result columns (type and name)") if the columns are defined to accept NULL values.  It works if the columns do not accept NULL values.

[sql3.zip](https://github.com/rbock/sqlpp11/files/521969/sql3.zip)

**Comment by rbock** [comment]
Got it! Thanks for the explanation.

If you feel adventurous, please try again. I pushed something without testing. I will add tests once I find a bit more time.

If the left hand side argument can be NULL, the right hand side is accepted even if it does not allow NULL.

See field_spec.h

**Comment by rbock** [comment]
Hi,

did you have a chance to test this? Are your use cases covered?

Best,

Roland

**Comment by kuriboshi** [comment]
The latest develop branch still has that final problem with the null/not null columns.  Otherwise everything works great for me.

**Comment by rbock** [comment]
OK, thanks, I'll look into that in the next few days.

**Comment by rbock** [comment]
Hi,

With the current develop, this works:

```C++
namespace greek
{
  SQLPP_ALIAS_PROVIDER(beta)
}

int main()
{
  // t.beta can be null, f.delta cannot
  static_assert(sqlpp::can_be_null_t<decltype(t.beta)>::value, "");
  static_assert(!sqlpp::can_be_null_t<decltype(f.delta)>::value, "");
  db(select(t.beta).from(t).unconditionally()
       .union_all(
       select(f.delta.as(greek::beta)).from(f).unconditionally()));
}
```

Cheers,

Roland

**Comment by rbock** [comment]
Released in 0.44.
